server/src/main/java/org/elasticsearch/index/mapper/IdLoader.java

```diff
@@ -66,21 +66,24 @@ final class TsIdLoader implements IdLoader {
         }
 
         public IdLoader.Leaf leaf(LeafStoredFieldLoader loader, LeafReader reader, int[] docIdsInLeaf) throws IOException {
-            IndexRouting.ExtractFromSource.Builder[] builders = new IndexRouting.ExtractFromSource.Builder[docIdsInLeaf.length];
-            for (int i = 0; i < builders.length; i++) {
-                builders[i] = indexRouting.builder();
-            }
+            IndexRouting.ExtractFromSource.Builder[] builders = null;
+            if (indexRouting != null) {
+                builders = new IndexRouting.ExtractFromSource.Builder[docIdsInLeaf.length];
+                for (int i = 0; i < builders.length; i++) {
+                    builders[i] = indexRouting.builder();
+                }
 
-            for (String routingField : routingPaths) {
-                // Routing field must always be keyword fields, so it is ok to use SortedSetDocValues directly here.
-                SortedSetDocValues dv = DocValues.getSortedSet(reader, routingField);
-                for (int i = 0; i < docIdsInLeaf.length; i++) {
-                    int docId = docIdsInLeaf[i];
-                    var builder = builders[i];
-                    if (dv.advanceExact(docId)) {
-                        for (int j = 0; j < dv.docValueCount(); j++) {
-                            BytesRef routingValue = dv.lookupOrd(dv.nextOrd());
-                            builder.addMatching(routingField, routingValue);
+                for (String routingField : routingPaths) {
```

Comment:
Can we just read the routing bit from the `routing_id` field? Or is this bwc logic?

Reply:
Yeah this is the old logic, for bwc.